jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
-
-                // Is the request fully consumed?
-                Throwable unconsumed = stream.consumeAvailable();
-                if (LOG.isDebugEnabled())
-                    LOG.debug("consumeAvailable: {} {} ", unconsumed == null, httpChannelState);
-
-                if (unconsumed != null && httpChannelState.getConnectionMetaData().isPersistent())
-                    failure = ExceptionUtil.combine(failure, unconsumed);
-
-                if (failure != null)
+                // If we are still not failing, is a last stream send needed or can we complete?
+                if (failure == null)
                 {
+                    doLastStreamSend = httpChannelState.lockedLastStreamSend();
+                    if (doLastStreamSend)
+                        response._writeCallback = httpChannelState._handlerInvoker;
+                    // or complete the stream if everything is done.
+                    else if (httpChannelState.lockedIsLastStreamSendCompleted())
+                        completeStream = httpChannelState._handled;
+                }
+                else
+                {
+                    // We are failing...
                     httpChannelState._callbackFailure = failure;
-                    if (!stream.isCommitted())
+
+                    // Can we and should we generate an error response?
+                    if (!stream.isCommitted() && !ExceptionUtil.isAssociated(failure, Request.Handler.AbortException.class))
+                    {
+                        // We are not committed, so we can send an error response.
                         errorResponse = new ErrorResponse(request);
-                    else
-                        completeStream = true;
+                    }
+                    else if (httpChannelState._handled)
+                    {
+                        // Callback and handling are completed, so it is just a matter of the last write
+                        if (httpChannelState.lockedIsLastStreamSendCompleted())
+                        {
+                            // We are committed, handling completed, and last write completed, so complete the stream now.
+                            completeStream = true;
+                        }
+                        else if (response.lockedIsWriting())
+                        {
+                            // We are currently writing, so let the completion of that write handle the failure
+                            httpChannelState._callbackFailure = failure;
+                            failedCallback = response._writeCallback;
+                            response._writeCallback = httpChannelState._handlerInvoker;
+                        }
+                        else
+                        {
+                            // There has been no last write, but we will just fail the stream instead.
+                            completeStream = true;
```

Comment:
Feels like it should be this, so the third leg is complete:


```suggestion
                            httpChannelState._streamSendState = StreamSendState.FAILED;
                            completeStream = true;
```